pages/[envId]/products/index.tsx

```diff
@@ -214,12 +217,31 @@ export const getStaticProps: GetStaticProps<Props, { envId: string }> = async co
     };
   }
 
-  const products = await getProductsForListing({ envId, previewApiKey }, !!context.preview);
-  const siteMenu = await getSiteMenu({ envId, previewApiKey }, !!context.preview);
+  const productsData = await getProductsForListing({ envId, previewApiKey }, !!context.preview);
+  const siteMenuData = await getSiteMenu({ envId, previewApiKey }, !!context.preview);
   const defaultMetadata = await getDefaultMetadata({ envId, previewApiKey }, !!context.preview);
 
+  if (!siteMenuData) {
+    throw new Error("Can't find the main menu item.")
+  }
+
+  let productsCircularReferences: ItemCircularReferenceMap = {};
+
+  const products = {
+    ...productsData,
+    items: productsData.items.map(product => {
+      const [sanitizedProduct, foundCycles] = sanitizeCircularData(product);
+      productsCircularReferences = {...productsCircularReferences, ...foundCycles};
+      return sanitizedProduct;

```

Comment:
Handling an array of products seems to be common enough to justify a utility function, it is quite a chunk of code.

Reply:
btw. it can be done functionally quite eloquently without a mutating object. 😝

Reply:
products don't actually have any linked item elements or rich text, so I removed the sanitation from the page. there are however potential linked items in article listing, so I added an extra method. I decided not to reuse the existing one because I felt it would add time complexity in its current state. some further refactor may be in order...